lib/ansible/plugins/strategy/__init__.py

```diff
@@ -1148,11 +1148,39 @@ def _evaluate_conditional(h):
                 skip_reason += ', not clearing facts and fact cache for %s' % target_host.name
         elif meta_action == 'clear_host_errors':
             if _evaluate_conditional(target_host):
+                missing_restore_state = []
                 for host in self._inventory.get_hosts(iterator._play.hosts):
                     self._tqm._failed_hosts.pop(host.name, False)
-                    self._tqm._unreachable_hosts.pop(host.name, False)
+                    unreachable = self._tqm._unreachable_hosts.pop(host.name, False)
+
+                    # If we're in a nested block, the failed host may not be in TQM's _failed_hosts yet.
+                    # Check is the host is failed with the PlayIterator since it considers child tasks.
+                    failed = iterator.is_failed(host)
+
                     iterator._host_states[host.name].fail_state = iterator.FAILED_NONE
+
+                    if not (failed or unreachable):
+                        continue
+
+                    # Strategy plugins should mark hosts as failed with PlayIterator.mark_host_failed.
+                    # If it hasn't, keep a list of hosts that can't be restored to add to msg.
+                    if failed and host.restore_state is None:
+                        missing_restore_state.append(host.name)
+                        continue
+
+                    if failed:
+                        iterator._host_states[host.name] = host.restore_state
+                        host.restore_state = None
+
+                    # Iterate over tasks until we get to clear_host_errors
+                    s, t = iterator.get_next_task_for_host(host, peek=False)
+                    while t is not None and t.args.get('_raw_params') != 'clear_host_errors':
+                        iterator._host_states[host.name].run_state = s.run_state
```

Comment:
`iterator.get_next_task_for_host` with `peek=False` already does `iterator._host_states[host.name] = s` so isn't this a noop?

Reply:
Yes, good catch!